In [25]:
from comet_ml import Experiment

In [26]:
import numpy as np
import pandas as pd

from tqdm import tqdm

import torch
import torch.nn as nn
import torch.optim as optim

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

import random

In [27]:
!pip install comet_ml

In [28]:
class CFG:

    api = "secret"
    project = "oskelly-mlp"
    workspace = "gp5_team1"
    num_workers = 2
    train_batch_size = 64
    test_batch_size = 512

    seed = 22
    comet = True

In [29]:
# Зафиксируем seed для воспроизводимости

def seed_everything(seed):
    random.seed(seed) # фиксируем генератор случайных чисел
    os.environ['PYTHONHASHSEED'] = str(seed) # фиксируем заполнения хешей
    np.random.seed(seed) # фиксируем генератор случайных чисел numpy
    torch.manual_seed(seed) # фиксируем генератор случайных чисел pytorch
    torch.cuda.manual_seed(seed) # фиксируем генератор случайных чисел для GPU
    #torch.backends.cudnn.deterministic = True # выбираем только детерминированные алгоритмы (для сверток)
    #torch.backends.cudnn.benchmark = False # фиксируем алгоритм вычисления сверток
seed_everything(CFG.seed)

In [30]:
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")

## Загрузим данные

данные (имеено цены) после этапа eda логарифмированы, будем учитывать это в работе и расчете метрик

In [31]:
X = np.load("X.npy")  # признаки
y = np.load("y.npy")  # целевая переменная цена

In [32]:
pd.Series(y).describe()

,0
count,9268.000000
mean,10.752261
std,0.802238
min,9.210040
25%,10.146796
50%,10.733588
75%,11.326295
max,12.611541


In [33]:
prices = np.exp(y)

pd.Series(prices).describe()

,0
count,9268.000000
mean,64446.734375
std,55786.210938
min,9996.997070
25%,25509.247070
50%,45870.992188
75%,82975.023438
max,300000.906250


Разделим данные на обучающую, валидационную и тестовую


In [34]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=CFG.seed
)

X_train, X_val, y_train, y_val = train_test_split(
    X_train,
    y_train,
    test_size=0.2,
    random_state=CFG.seed
)

Создадим датасеты и загрузчики данных в формате PyTorch.

В отличие от примера из семинара, здесь мы не загружаем готовый датасет через `datasets.MNIST`, потому что признаки и целевая переменная уже подготовлены заранее и загружены из файлов `X.npy` и `y.npy`.

Поэтому сначала переводим массивы NumPy в тензоры, затем создаём `TensorDataset` и `DataLoader`.

In [35]:
X_train_tensor = torch.tensor(X_train, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train, dtype=torch.float32)

X_val_tensor = torch.tensor(X_val, dtype=torch.float32)
y_val_tensor = torch.tensor(y_val, dtype=torch.float32)

X_test_tensor = torch.tensor(X_test, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test, dtype=torch.float32)

In [36]:
train_data = torch.utils.data.TensorDataset(X_train_tensor, y_train_tensor)
val_data = torch.utils.data.TensorDataset(X_val_tensor, y_val_tensor)
test_data = torch.utils.data.TensorDataset(X_test_tensor, y_test_tensor)

In [37]:
train_loader = torch.utils.data.DataLoader(train_data, batch_size=CFG.train_batch_size, shuffle=True, num_workers=CFG.num_workers)

val_loader = torch.utils.data.DataLoader(val_data, batch_size=CFG.test_batch_size, shuffle=False, num_workers=CFG.num_workers)

test_loader = torch.utils.data.DataLoader(test_data, batch_size=CFG.test_batch_size, shuffle=False, num_workers=CFG.num_workers)

напишем простую MLP-модель - сейчас она будет состоять из двух скрытых полносвязных слоёв и одного выходного слоя, который предсказывает одно число — цену товара

In [38]:
class PriceMLP(nn.Module):
    def __init__(self, input_dim, hidden_1, hidden_2, activation):
        super().__init__()

        self.fc1 = nn.Linear(input_dim, hidden_1)
        self.fc2 = nn.Linear(hidden_1, hidden_2)
        self.fc3 = nn.Linear(hidden_2, 1)

        if activation == "relu":
            self.activation = nn.ReLU()
        elif activation == "tanh":
            self.activation = nn.Tanh()

    def forward(self, x):
        x = self.activation(self.fc1(x))
        x = self.activation(self.fc2(x))
        x = self.fc3(x)
        return x.squeeze()

напишем теперь функции для обучения модели

In [39]:
def train(model, device, train_loader, optimizer, criterion, epoch, logger=None):
    model.train()
    train_loss = 0

    for data, target in tqdm(train_loader):
        data = data.to(device)
        target = target.to(device)

        optimizer.zero_grad()

        output = model(data)
        loss = criterion(output, target)

        loss.backward()
        optimizer.step()

        train_loss += loss.item()

    train_loss = train_loss / len(train_loader)

    print(f"Epoch {epoch}: train_loss = {train_loss:.4f}")

    if logger:
        logger.log_metric("train_loss", train_loss, step=epoch)

    return train_loss

In [40]:
def test(model, device, loader, criterion, epoch, name="val_loss", logger=None):
    model.eval()
    test_loss = 0

    with torch.no_grad():
        for data, target in tqdm(loader):
            data = data.to(device)
            target = target.to(device)

            output = model(data)
            loss = criterion(output, target)

            test_loss += loss.item()

    test_loss = test_loss / len(loader)

    print(f"Epoch {epoch}: {name} = {test_loss:.4f}")

    if logger:
        logger.log_metric(name, test_loss, step=epoch)

    return test_loss

После обучения модели нужно получить предсказания на тестовой выборке

Модель предсказывает логарифм цены, поэтому перед расчётом метрик переводим обратно в рубли

В качестве метрик используем MAE, RMSE, MAPE и R квадрат

In [41]:
def get_predictions(model, device, loader):
    model.eval()

    y_pred = []
    y_true = []

    with torch.no_grad():
        for data, target in tqdm(loader):
            data = data.to(device)

            output = model(data)

            y_pred.extend(output.cpu().numpy())
            y_true.extend(target.numpy())

    y_pred = np.array(y_pred)
    y_true = np.array(y_true)

    return y_true, y_pred

In [42]:
def calculate_price_metrics(y_true_log, y_pred_log):
    y_true_price = np.exp(y_true_log)
    y_pred_price = np.exp(y_pred_log)

    mae = mean_absolute_error(y_true_price, y_pred_price)
    rmse = mean_squared_error(y_true_price, y_pred_price) ** 0.5
    mape = np.mean(np.abs((y_true_price - y_pred_price) / y_true_price)) * 100
    r2 = r2_score(y_true_price, y_pred_price)

    return mae, rmse, mape, r2

создадим массив, гда зададим разные эксперименты, что будем менять - число нейронов в скрытых слоях, функции активации, оптимизаторы и learning rate


In [43]:
experiments = [
    {
        "name": "128-64_relu_adam_lr0001_10ep",
        "hidden_1": 128,
        "hidden_2": 64,
        "activation": "relu",
        "optimizer": optim.Adam,
        "lr": 0.001,
        "epochs": 10
    },
    {
        "name": "256-128_relu_adam_lr0001_10ep",
        "hidden_1": 256,
        "hidden_2": 128,
        "activation": "relu",
        "optimizer": optim.Adam,
        "lr": 0.001,
        "epochs": 10
    },
    {
        "name": "512-256_relu_adam_lr0001_10ep",
        "hidden_1": 512,
        "hidden_2": 256,
        "activation": "relu",
        "optimizer": optim.Adam,
        "lr": 0.001,
        "epochs": 10
    },
    {
        "name": "512-256_relu_adam_lr0005_14ep",
        "hidden_1": 512,
        "hidden_2": 256,
        "activation": "relu",
        "optimizer": optim.Adam,
        "lr": 0.0005,
        "epochs": 14
    },
    {
        "name": "768-384_relu_adam_lr0005_14ep",
        "hidden_1": 768,
        "hidden_2": 384,
        "activation": "relu",
        "optimizer": optim.Adam,
        "lr": 0.0005,
        "epochs": 14
    },

    {
        "name": "256-128_tanh_adam_lr0001_10ep",
        "hidden_1": 256,
        "hidden_2": 128,
        "activation": "tanh",
        "optimizer": optim.Adam,
        "lr": 0.001,
        "epochs": 10
    },
    {
        "name": "512-256_tanh_adam_lr0005_14ep",
        "hidden_1": 512,
        "hidden_2": 256,
        "activation": "tanh",
        "optimizer": optim.Adam,
        "lr": 0.0005,
        "epochs": 14
    },
    {
        "name": "768-384_tanh_adam_lr0005_14ep",
        "hidden_1": 768,
        "hidden_2": 384,
        "activation": "tanh",
        "optimizer": optim.Adam,
        "lr": 0.0005,
        "epochs": 14
    },

    {
        "name": "256-128_relu_sgd_lr001_10ep",
        "hidden_1": 256,
        "hidden_2": 128,
        "activation": "relu",
        "optimizer": optim.SGD,
        "lr": 0.01,
        "epochs": 10
    },
    {
        "name": "512-256_relu_sgd_lr001_14ep",
        "hidden_1": 512,
        "hidden_2": 256,
        "activation": "relu",
        "optimizer": optim.SGD,
        "lr": 0.01,
        "epochs": 14
    },
    {
        "name": "256-128_tanh_sgd_lr001_10ep",
        "hidden_1": 256,
        "hidden_2": 128,
        "activation": "tanh",
        "optimizer": optim.SGD,
        "lr": 0.01,
        "epochs": 10
    },
    {
        "name": "512-256_tanh_sgd_lr001_14ep",
        "hidden_1": 512,
        "hidden_2": 256,
        "activation": "tanh",
        "optimizer": optim.SGD,
        "lr": 0.01,
        "epochs": 14
    },
]

In [44]:
def run_experiment(exp):
    seed_everything(CFG.seed)

    logger = None

    if CFG.comet:
        logger = Experiment(api_key=CFG.api, project_name=CFG.project, workspace=CFG.workspace)

        logger.set_name(exp["name"])

        logger.log_parameters({
            "hidden_1": exp["hidden_1"],
            "hidden_2": exp["hidden_2"],
            "activation": exp["activation"],
            "optimizer": exp["optimizer"].__name__,
            "lr": exp["lr"],
            "epochs": exp["epochs"],
            "batch_size": CFG.train_batch_size,
            "loss": "MSELoss",
            "seed": CFG.seed
        })

    model = PriceMLP(X_train.shape[1], exp["hidden_1"], exp["hidden_2"], exp["activation"]).to(device)

    criterion = nn.MSELoss()
    optimizer = exp["optimizer"](model.parameters(), lr=exp["lr"])

    for epoch in range(1, exp["epochs"] + 1):
        train(model, device, train_loader, optimizer, criterion, epoch, logger)
        test(model, device, val_loader, criterion, epoch, "val_loss", logger)

    y_true, y_pred = get_predictions(model, device, test_loader)
    mae, rmse, mape, r2 = calculate_price_metrics(y_true, y_pred)

    result = {
        "name": exp["name"],
        "hidden_1": exp["hidden_1"],
        "hidden_2": exp["hidden_2"],
        "activation": exp["activation"],
        "optimizer": exp["optimizer"].__name__,
        "lr": exp["lr"],
        "epochs": exp["epochs"],
        "MAE": mae,
        "RMSE": rmse,
        "MAPE": mape,
        "R2": r2
    }

    if logger:
        logger.log_metrics({"MAE": mae, "RMSE": rmse, "MAPE": mape, "R2": r2})

        model_path = exp["name"] + ".pt"
        torch.save(model.state_dict(), model_path)

        logger.log_model("Light-MLP", model_path)
        logger.end()

    return result, model

In [45]:
results = []
models = {}

for exp in experiments:
    result, model = run_experiment(exp)
    results.append(result)
    models[exp["name"]] = model

results_df = pd.DataFrame(results)

COMET WARNING: To get all data logged automatically, import comet_ml before the following modules: sklearn, torch.
COMET WARNING: As you are running in a Jupyter environment, you will need to call `experiment.end()` when finished to ensure all metrics and code are logged before exiting.
COMET INFO: Experiment is live on comet.com https://www.comet.com/gp5-team1/oskelly-mlp/05cac8b1bcb04eaea1f54d522fc02c70

COMET INFO: Couldn't find a Git repository in '/content' nor in any parent directory. Set `COMET_GIT_DIRECTORY` if your Git Repository is elsewhere.
100%|██████████| 93/93 [00:00<00:00, 104.76it/s]


Epoch 1: train_loss = 25.9192


100%|██████████| 3/3 [00:00<00:00,  9.33it/s]


Epoch 1: val_loss = 1.7699


100%|██████████| 93/93 [00:01<00:00, 76.33it/s]


Epoch 2: train_loss = 1.1803


100%|██████████| 3/3 [00:00<00:00, 11.27it/s]


Epoch 2: val_loss = 0.9264


100%|██████████| 93/93 [00:01<00:00, 75.48it/s]


Epoch 3: train_loss = 0.6530


100%|██████████| 3/3 [00:00<00:00,  9.01it/s]


Epoch 3: val_loss = 0.5722


100%|██████████| 93/93 [00:01<00:00, 73.66it/s]


Epoch 4: train_loss = 0.4350


100%|██████████| 3/3 [00:00<00:00, 11.36it/s]


Epoch 4: val_loss = 0.4444


100%|██████████| 93/93 [00:01<00:00, 58.26it/s]


Epoch 5: train_loss = 0.3479


100%|██████████| 3/3 [00:00<00:00,  5.02it/s]


Epoch 5: val_loss = 0.3849


100%|██████████| 93/93 [00:01<00:00, 49.19it/s]


Epoch 6: train_loss = 0.3020


100%|██████████| 3/3 [00:00<00:00,  6.42it/s]


Epoch 6: val_loss = 0.3676


100%|██████████| 93/93 [00:01<00:00, 48.03it/s]


Epoch 7: train_loss = 0.2693


100%|██████████| 3/3 [00:00<00:00,  9.72it/s]


Epoch 7: val_loss = 0.3442


100%|██████████| 93/93 [00:01<00:00, 60.21it/s]


Epoch 8: train_loss = 0.2456


100%|██████████| 3/3 [00:00<00:00, 11.50it/s]


Epoch 8: val_loss = 0.3260


100%|██████████| 93/93 [00:00<00:00, 102.72it/s]


Epoch 9: train_loss = 0.2244


100%|██████████| 3/3 [00:00<00:00, 25.78it/s]


Epoch 9: val_loss = 0.3201


100%|██████████| 93/93 [00:00<00:00, 151.88it/s]


Epoch 10: train_loss = 0.2217


100%|██████████| 3/3 [00:00<00:00, 23.83it/s]


Epoch 10: val_loss = 0.3275


100%|██████████| 4/4 [00:00<00:00, 28.71it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 128-64_relu_adam_lr0001_10ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/05cac8b1bcb04eaea1f54d522fc02c70
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 29941.181640625
COMET INFO:     MAPE            : 60.24631118774414
COMET INFO:     R2              : 0.12788057327270508
COMET INFO:     RMSE            : 51144.44986506356
COMET INFO:     train_loss [10] : (0.221749792256022, 25.919230373956825)
COMET INFO:     val_loss [10]   : (0.32010794679323834, 1.7698566913604736)
COMET INFO:   Others:
COMET INFO:     Name         : 128-64_relu_a

Epoch 1: train_loss = 18.5613


100%|██████████| 3/3 [00:00<00:00, 21.88it/s]


Epoch 1: val_loss = 1.4295


100%|██████████| 93/93 [00:00<00:00, 120.26it/s]


Epoch 2: train_loss = 0.9128


100%|██████████| 3/3 [00:00<00:00, 23.53it/s]


Epoch 2: val_loss = 0.6986


100%|██████████| 93/93 [00:00<00:00, 119.27it/s]


Epoch 3: train_loss = 0.5020


100%|██████████| 3/3 [00:00<00:00, 22.09it/s]


Epoch 3: val_loss = 0.4596


100%|██████████| 93/93 [00:00<00:00, 102.92it/s]


Epoch 4: train_loss = 0.3457


100%|██████████| 3/3 [00:00<00:00, 14.31it/s]


Epoch 4: val_loss = 0.4003


100%|██████████| 93/93 [00:01<00:00, 88.23it/s] 


Epoch 5: train_loss = 0.2884


100%|██████████| 3/3 [00:00<00:00, 15.02it/s]


Epoch 5: val_loss = 0.3589


100%|██████████| 93/93 [00:01<00:00, 85.69it/s]


Epoch 6: train_loss = 0.2453


100%|██████████| 3/3 [00:00<00:00, 21.85it/s]


Epoch 6: val_loss = 0.3284


100%|██████████| 93/93 [00:00<00:00, 129.76it/s]


Epoch 7: train_loss = 0.2140


100%|██████████| 3/3 [00:00<00:00, 24.01it/s]


Epoch 7: val_loss = 0.3044


100%|██████████| 93/93 [00:00<00:00, 120.21it/s]


Epoch 8: train_loss = 0.1909


100%|██████████| 3/3 [00:00<00:00, 22.42it/s]


Epoch 8: val_loss = 0.2925


100%|██████████| 93/93 [00:00<00:00, 113.53it/s]


Epoch 9: train_loss = 0.1757


100%|██████████| 3/3 [00:00<00:00, 22.18it/s]


Epoch 9: val_loss = 0.2910


100%|██████████| 93/93 [00:00<00:00, 111.47it/s]


Epoch 10: train_loss = 0.1682


100%|██████████| 3/3 [00:00<00:00, 24.03it/s]


Epoch 10: val_loss = 0.3043


100%|██████████| 4/4 [00:00<00:00, 26.85it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 256-128_relu_adam_lr0001_10ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/982d076b9024451a9bece446468ede8a
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 28249.083984375
COMET INFO:     MAPE            : 55.04450988769531
COMET INFO:     R2              : 0.140946626663208
COMET INFO:     RMSE            : 50759.88085092399
COMET INFO:     train_loss [10] : (0.16821870464150623, 18.56129199586889)
COMET INFO:     val_loss [10]   : (0.2909564773241679, 1.4295318921407063)
COMET INFO:   Others:
COMET INFO:     Name         : 256-128_relu_a

Epoch 1: train_loss = 13.0375


100%|██████████| 3/3 [00:00<00:00, 19.74it/s]


Epoch 1: val_loss = 1.0635


100%|██████████| 93/93 [00:01<00:00, 89.60it/s] 


Epoch 2: train_loss = 0.6987


100%|██████████| 3/3 [00:00<00:00, 20.09it/s]


Epoch 2: val_loss = 0.5429


100%|██████████| 93/93 [00:01<00:00, 66.70it/s]


Epoch 3: train_loss = 0.4031


100%|██████████| 3/3 [00:00<00:00, 12.35it/s]


Epoch 3: val_loss = 0.4621


100%|██████████| 93/93 [00:01<00:00, 61.61it/s]


Epoch 4: train_loss = 0.3120


100%|██████████| 3/3 [00:00<00:00, 12.21it/s]


Epoch 4: val_loss = 0.3772


100%|██████████| 93/93 [00:01<00:00, 82.39it/s]


Epoch 5: train_loss = 0.2498


100%|██████████| 3/3 [00:00<00:00, 20.33it/s]


Epoch 5: val_loss = 0.3306


100%|██████████| 93/93 [00:01<00:00, 88.89it/s]


Epoch 6: train_loss = 0.2096


100%|██████████| 3/3 [00:00<00:00, 19.64it/s]


Epoch 6: val_loss = 0.3021


100%|██████████| 93/93 [00:01<00:00, 88.40it/s]


Epoch 7: train_loss = 0.1970


100%|██████████| 3/3 [00:00<00:00, 19.79it/s]


Epoch 7: val_loss = 0.2969


100%|██████████| 93/93 [00:01<00:00, 76.53it/s]


Epoch 8: train_loss = 0.1683


100%|██████████| 3/3 [00:00<00:00, 19.02it/s]


Epoch 8: val_loss = 0.2804


100%|██████████| 93/93 [00:01<00:00, 69.16it/s]


Epoch 9: train_loss = 0.1512


100%|██████████| 3/3 [00:00<00:00, 20.35it/s]


Epoch 9: val_loss = 0.2840


100%|██████████| 93/93 [00:01<00:00, 75.35it/s]


Epoch 10: train_loss = 0.1359


100%|██████████| 3/3 [00:00<00:00, 19.24it/s]


Epoch 10: val_loss = 0.2778


100%|██████████| 4/4 [00:00<00:00, 24.38it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 512-256_relu_adam_lr0001_10ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/b874850dbb7f4216b71ef6e03afa0d62
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 24753.5234375
COMET INFO:     MAPE            : 40.47548294067383
COMET INFO:     R2              : 0.38621097803115845
COMET INFO:     RMSE            : 42906.20393369705
COMET INFO:     train_loss [10] : (0.1359355265414843, 13.037479999885765)
COMET INFO:     val_loss [10]   : (0.27777086695035297, 1.0635415315628052)
COMET INFO:   Others:
COMET INFO:     Name         : 512-256_relu_

Epoch 1: train_loss = 20.6135


100%|██████████| 3/3 [00:00<00:00, 12.58it/s]


Epoch 1: val_loss = 1.4488


100%|██████████| 93/93 [00:01<00:00, 74.88it/s]


Epoch 2: train_loss = 0.9372


100%|██████████| 3/3 [00:00<00:00, 17.59it/s]


Epoch 2: val_loss = 0.7283


100%|██████████| 93/93 [00:01<00:00, 87.81it/s]


Epoch 3: train_loss = 0.5138


100%|██████████| 3/3 [00:00<00:00, 19.82it/s]


Epoch 3: val_loss = 0.4966


100%|██████████| 93/93 [00:01<00:00, 88.08it/s]


Epoch 4: train_loss = 0.3668


100%|██████████| 3/3 [00:00<00:00, 19.84it/s]


Epoch 4: val_loss = 0.3984


100%|██████████| 93/93 [00:01<00:00, 86.27it/s]


Epoch 5: train_loss = 0.2964


100%|██████████| 3/3 [00:00<00:00, 21.06it/s]


Epoch 5: val_loss = 0.3485


100%|██████████| 93/93 [00:01<00:00, 88.84it/s]


Epoch 6: train_loss = 0.2507


100%|██████████| 3/3 [00:00<00:00, 19.19it/s]


Epoch 6: val_loss = 0.3198


100%|██████████| 93/93 [00:01<00:00, 88.64it/s]


Epoch 7: train_loss = 0.2232


100%|██████████| 3/3 [00:00<00:00, 18.80it/s]


Epoch 7: val_loss = 0.3047


100%|██████████| 93/93 [00:01<00:00, 77.07it/s]


Epoch 8: train_loss = 0.1956


100%|██████████| 3/3 [00:00<00:00, 19.41it/s]


Epoch 8: val_loss = 0.3181


100%|██████████| 93/93 [00:01<00:00, 74.30it/s]


Epoch 9: train_loss = 0.1762


100%|██████████| 3/3 [00:00<00:00, 19.47it/s]


Epoch 9: val_loss = 0.2873


100%|██████████| 93/93 [00:01<00:00, 59.82it/s]


Epoch 10: train_loss = 0.1631


100%|██████████| 3/3 [00:00<00:00, 11.71it/s]


Epoch 10: val_loss = 0.2918


100%|██████████| 93/93 [00:01<00:00, 57.16it/s]


Epoch 11: train_loss = 0.1524


100%|██████████| 3/3 [00:00<00:00, 18.82it/s]


Epoch 11: val_loss = 0.3135


100%|██████████| 93/93 [00:01<00:00, 74.55it/s]


Epoch 12: train_loss = 0.1410


100%|██████████| 3/3 [00:00<00:00, 19.70it/s]


Epoch 12: val_loss = 0.2621


100%|██████████| 93/93 [00:01<00:00, 78.49it/s]


Epoch 13: train_loss = 0.1336


100%|██████████| 3/3 [00:00<00:00, 18.94it/s]


Epoch 13: val_loss = 0.2846


100%|██████████| 93/93 [00:01<00:00, 77.35it/s]


Epoch 14: train_loss = 0.1190


100%|██████████| 3/3 [00:00<00:00, 19.81it/s]


Epoch 14: val_loss = 0.2535


100%|██████████| 4/4 [00:00<00:00, 21.75it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 512-256_relu_adam_lr0005_14ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/0c2b3c8ebe064354a0781698fb87110b
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 23981.755859375
COMET INFO:     MAPE            : 41.85343933105469
COMET INFO:     R2              : 0.41850143671035767
COMET INFO:     RMSE            : 41762.34169679665
COMET INFO:     train_loss [14] : (0.11901561691555926, 20.61345573266347)
COMET INFO:     val_loss [14]   : (0.253483384847641, 1.4488027493158977)
COMET INFO:   Others:
COMET INFO:     Name         : 512-256_relu_

Epoch 1: train_loss = 16.1072


100%|██████████| 3/3 [00:00<00:00, 16.18it/s]


Epoch 1: val_loss = 1.3259


100%|██████████| 93/93 [00:01<00:00, 58.97it/s]


Epoch 2: train_loss = 0.8109


100%|██████████| 3/3 [00:00<00:00, 11.01it/s]


Epoch 2: val_loss = 0.6525


100%|██████████| 93/93 [00:02<00:00, 44.33it/s]


Epoch 3: train_loss = 0.4486


100%|██████████| 3/3 [00:00<00:00, 10.29it/s]


Epoch 3: val_loss = 0.4283


100%|██████████| 93/93 [00:01<00:00, 60.28it/s]


Epoch 4: train_loss = 0.3211


100%|██████████| 3/3 [00:00<00:00, 16.72it/s]


Epoch 4: val_loss = 0.3768


100%|██████████| 93/93 [00:01<00:00, 62.37it/s]


Epoch 5: train_loss = 0.2608


100%|██████████| 3/3 [00:00<00:00, 17.22it/s]


Epoch 5: val_loss = 0.3844


100%|██████████| 93/93 [00:01<00:00, 64.05it/s]


Epoch 6: train_loss = 0.2245


100%|██████████| 3/3 [00:00<00:00, 17.54it/s]


Epoch 6: val_loss = 0.3301


100%|██████████| 93/93 [00:01<00:00, 62.11it/s]


Epoch 7: train_loss = 0.1880


100%|██████████| 3/3 [00:00<00:00, 17.57it/s]


Epoch 7: val_loss = 0.2997


100%|██████████| 93/93 [00:01<00:00, 55.33it/s]


Epoch 8: train_loss = 0.1663


100%|██████████| 3/3 [00:00<00:00, 17.05it/s]


Epoch 8: val_loss = 0.2926


100%|██████████| 93/93 [00:01<00:00, 48.14it/s]


Epoch 9: train_loss = 0.1463


100%|██████████| 3/3 [00:00<00:00, 11.08it/s]


Epoch 9: val_loss = 0.2977


100%|██████████| 93/93 [00:02<00:00, 38.65it/s]


Epoch 10: train_loss = 0.1357


100%|██████████| 3/3 [00:00<00:00, 12.66it/s]


Epoch 10: val_loss = 0.3036


100%|██████████| 93/93 [00:01<00:00, 53.69it/s]


Epoch 11: train_loss = 0.1203


100%|██████████| 3/3 [00:00<00:00, 16.69it/s]


Epoch 11: val_loss = 0.2771


100%|██████████| 93/93 [00:01<00:00, 53.98it/s]


Epoch 12: train_loss = 0.1167


100%|██████████| 3/3 [00:00<00:00, 17.31it/s]


Epoch 12: val_loss = 0.2862


100%|██████████| 93/93 [00:01<00:00, 53.68it/s]


Epoch 13: train_loss = 0.1013


100%|██████████| 3/3 [00:00<00:00, 17.71it/s]


Epoch 13: val_loss = 0.2726


100%|██████████| 93/93 [00:01<00:00, 53.09it/s]


Epoch 14: train_loss = 0.0922


100%|██████████| 3/3 [00:00<00:00, 16.68it/s]


Epoch 14: val_loss = 0.2662


100%|██████████| 4/4 [00:00<00:00, 20.19it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 768-384_relu_adam_lr0005_14ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/31535e75320d4e348c39f2e1d39cab69
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 25551.587890625
COMET INFO:     MAPE            : 47.026695251464844
COMET INFO:     R2              : 0.31434935331344604
COMET INFO:     RMSE            : 45348.40204461454
COMET INFO:     train_loss [14] : (0.09216774094809768, 16.10719450763477)
COMET INFO:     val_loss [14]   : (0.266197403271993, 1.3258723417917888)
COMET INFO:   Others:
COMET INFO:     Name         : 768-384_relu

Epoch 1: train_loss = 11.7517


100%|██████████| 3/3 [00:00<00:00, 20.04it/s]


Epoch 1: val_loss = 0.6492


100%|██████████| 93/93 [00:00<00:00, 127.07it/s]


Epoch 2: train_loss = 0.6441


100%|██████████| 3/3 [00:00<00:00, 21.20it/s]


Epoch 2: val_loss = 0.6512


100%|██████████| 93/93 [00:00<00:00, 120.57it/s]


Epoch 3: train_loss = 0.6437


100%|██████████| 3/3 [00:00<00:00, 21.85it/s]


Epoch 3: val_loss = 0.6486


100%|██████████| 93/93 [00:00<00:00, 120.86it/s]


Epoch 4: train_loss = 0.6427


100%|██████████| 3/3 [00:00<00:00, 21.04it/s]


Epoch 4: val_loss = 0.6530


100%|██████████| 93/93 [00:00<00:00, 125.81it/s]


Epoch 5: train_loss = 0.6441


100%|██████████| 3/3 [00:00<00:00, 21.55it/s]


Epoch 5: val_loss = 0.6487


100%|██████████| 93/93 [00:00<00:00, 120.30it/s]


Epoch 6: train_loss = 0.6427


100%|██████████| 3/3 [00:00<00:00, 20.27it/s]


Epoch 6: val_loss = 0.6531


100%|██████████| 93/93 [00:00<00:00, 122.17it/s]


Epoch 7: train_loss = 0.6441


100%|██████████| 3/3 [00:00<00:00, 21.32it/s]


Epoch 7: val_loss = 0.6476


100%|██████████| 93/93 [00:00<00:00, 118.97it/s]


Epoch 8: train_loss = 0.6430


100%|██████████| 3/3 [00:00<00:00, 22.23it/s]


Epoch 8: val_loss = 0.6473


100%|██████████| 93/93 [00:00<00:00, 124.58it/s]


Epoch 9: train_loss = 0.6447


100%|██████████| 3/3 [00:00<00:00, 19.83it/s]


Epoch 9: val_loss = 0.6476


100%|██████████| 93/93 [00:00<00:00, 127.61it/s]


Epoch 10: train_loss = 0.6417


100%|██████████| 3/3 [00:00<00:00, 22.17it/s]


Epoch 10: val_loss = 0.6468


100%|██████████| 4/4 [00:00<00:00, 25.59it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 256-128_tanh_adam_lr0001_10ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/114286a653b54cbb9b4fed98dd98a61e
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 37926.5703125
COMET INFO:     MAPE            : 80.68408966064453
COMET INFO:     R2              : -0.0900810956954956
COMET INFO:     RMSE            : 57179.44693681463
COMET INFO:     train_loss [10] : (0.64173789338399, 11.751715081673796)
COMET INFO:     val_loss [10]   : (0.6468304594357809, 0.6531474788983663)
COMET INFO:   Others:
COMET INFO:     Name         : 256-128_tanh_ada

Epoch 1: train_loss = 11.9553


100%|██████████| 3/3 [00:00<00:00, 19.40it/s]


Epoch 1: val_loss = 0.6821


100%|██████████| 93/93 [00:01<00:00, 84.84it/s]


Epoch 2: train_loss = 0.6606


100%|██████████| 3/3 [00:00<00:00, 19.32it/s]


Epoch 2: val_loss = 0.6702


100%|██████████| 93/93 [00:01<00:00, 87.19it/s]


Epoch 3: train_loss = 0.6533


100%|██████████| 3/3 [00:00<00:00, 17.28it/s]


Epoch 3: val_loss = 0.6626


100%|██████████| 93/93 [00:01<00:00, 84.04it/s]


Epoch 4: train_loss = 0.6488


100%|██████████| 3/3 [00:00<00:00, 18.65it/s]


Epoch 4: val_loss = 0.6577


100%|██████████| 93/93 [00:01<00:00, 87.32it/s]


Epoch 5: train_loss = 0.6443


100%|██████████| 3/3 [00:00<00:00, 19.16it/s]


Epoch 5: val_loss = 0.6533


100%|██████████| 93/93 [00:01<00:00, 84.00it/s]


Epoch 6: train_loss = 0.6432


100%|██████████| 3/3 [00:00<00:00, 18.77it/s]


Epoch 6: val_loss = 0.6524


100%|██████████| 93/93 [00:01<00:00, 86.94it/s]


Epoch 7: train_loss = 0.6426


100%|██████████| 3/3 [00:00<00:00, 18.79it/s]


Epoch 7: val_loss = 0.6496


100%|██████████| 93/93 [00:01<00:00, 85.67it/s]


Epoch 8: train_loss = 0.6398


100%|██████████| 3/3 [00:00<00:00, 17.81it/s]


Epoch 8: val_loss = 0.6456


100%|██████████| 93/93 [00:01<00:00, 74.94it/s]


Epoch 9: train_loss = 0.6353


100%|██████████| 3/3 [00:00<00:00, 11.74it/s]


Epoch 9: val_loss = 0.6545


100%|██████████| 93/93 [00:01<00:00, 61.34it/s]


Epoch 10: train_loss = 0.6341


100%|██████████| 3/3 [00:00<00:00, 11.03it/s]


Epoch 10: val_loss = 0.6389


100%|██████████| 93/93 [00:01<00:00, 68.82it/s]


Epoch 11: train_loss = 0.6282


100%|██████████| 3/3 [00:00<00:00, 18.63it/s]


Epoch 11: val_loss = 0.6313


100%|██████████| 93/93 [00:01<00:00, 87.61it/s]


Epoch 12: train_loss = 0.6138


100%|██████████| 3/3 [00:00<00:00, 18.57it/s]


Epoch 12: val_loss = 0.6050


100%|██████████| 93/93 [00:01<00:00, 85.03it/s]


Epoch 13: train_loss = 0.5474


100%|██████████| 3/3 [00:00<00:00, 17.60it/s]


Epoch 13: val_loss = 0.4948


100%|██████████| 93/93 [00:01<00:00, 85.80it/s]


Epoch 14: train_loss = 0.4014


100%|██████████| 3/3 [00:00<00:00, 18.72it/s]


Epoch 14: val_loss = 0.3514


100%|██████████| 4/4 [00:00<00:00, 23.09it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 512-256_tanh_adam_lr0005_14ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/1979ca664e2948059cb7af0876fb8376
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 26925.169921875
COMET INFO:     MAPE            : 53.47491455078125
COMET INFO:     R2              : 0.3838074207305908
COMET INFO:     RMSE            : 42990.130216132166
COMET INFO:     train_loss [14] : (0.40142604420261996, 11.955341323729485)
COMET INFO:     val_loss [14]   : (0.3513543903827667, 0.6820870240529379)
COMET INFO:   Others:
COMET INFO:     Name         : 512-256_tan

Epoch 1: train_loss = 9.2445


100%|██████████| 3/3 [00:00<00:00, 16.69it/s]


Epoch 1: val_loss = 0.6847


100%|██████████| 93/93 [00:01<00:00, 49.05it/s]


Epoch 2: train_loss = 0.6601


100%|██████████| 3/3 [00:00<00:00, 10.82it/s]


Epoch 2: val_loss = 0.6705


100%|██████████| 93/93 [00:01<00:00, 48.96it/s]


Epoch 3: train_loss = 0.6503


100%|██████████| 3/3 [00:00<00:00, 15.90it/s]


Epoch 3: val_loss = 0.6590


100%|██████████| 93/93 [00:01<00:00, 63.65it/s]


Epoch 4: train_loss = 0.6485


100%|██████████| 3/3 [00:00<00:00, 16.50it/s]


Epoch 4: val_loss = 0.6529


100%|██████████| 93/93 [00:01<00:00, 64.40it/s]


Epoch 5: train_loss = 0.6412


100%|██████████| 3/3 [00:00<00:00, 15.91it/s]


Epoch 5: val_loss = 0.6439


100%|██████████| 93/93 [00:01<00:00, 59.82it/s]


Epoch 6: train_loss = 0.6244


100%|██████████| 3/3 [00:00<00:00, 15.06it/s]


Epoch 6: val_loss = 0.6251


100%|██████████| 93/93 [00:01<00:00, 59.02it/s]


Epoch 7: train_loss = 0.5903


100%|██████████| 3/3 [00:00<00:00, 16.53it/s]


Epoch 7: val_loss = 0.5787


100%|██████████| 93/93 [00:01<00:00, 60.56it/s]


Epoch 8: train_loss = 0.5390


100%|██████████| 3/3 [00:00<00:00, 15.84it/s]


Epoch 8: val_loss = 0.5225


100%|██████████| 93/93 [00:01<00:00, 48.74it/s]


Epoch 9: train_loss = 0.4762


100%|██████████| 3/3 [00:00<00:00,  9.57it/s]


Epoch 9: val_loss = 0.4518


100%|██████████| 93/93 [00:02<00:00, 45.47it/s]


Epoch 10: train_loss = 0.3849


100%|██████████| 3/3 [00:00<00:00, 13.66it/s]


Epoch 10: val_loss = 0.3762


100%|██████████| 93/93 [00:01<00:00, 61.40it/s]


Epoch 11: train_loss = 0.3076


100%|██████████| 3/3 [00:00<00:00, 16.70it/s]


Epoch 11: val_loss = 0.3091


100%|██████████| 93/93 [00:01<00:00, 59.88it/s]


Epoch 12: train_loss = 0.2516


100%|██████████| 3/3 [00:00<00:00, 16.12it/s]


Epoch 12: val_loss = 0.2689


100%|██████████| 93/93 [00:01<00:00, 62.39it/s]


Epoch 13: train_loss = 0.2304


100%|██████████| 3/3 [00:00<00:00, 17.24it/s]


Epoch 13: val_loss = 0.2563


100%|██████████| 93/93 [00:01<00:00, 62.19it/s]


Epoch 14: train_loss = 0.2176


100%|██████████| 3/3 [00:00<00:00, 16.03it/s]


Epoch 14: val_loss = 0.2508


100%|██████████| 4/4 [00:00<00:00, 19.76it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 768-384_tanh_adam_lr0005_14ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/a5e1ec268f334aeab26da7ffc614b6c2
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 22951.03515625
COMET INFO:     MAPE            : 44.24778747558594
COMET INFO:     R2              : 0.5384289026260376
COMET INFO:     RMSE            : 37207.44183627786
COMET INFO:     train_loss [14] : (0.2176120747481623, 9.244477340290624)
COMET INFO:     val_loss [14]   : (0.25083182752132416, 0.6846959988276163)
COMET INFO:   Others:
COMET INFO:     Name         : 768-384_tanh_a

Epoch 1: train_loss = 10.7231


100%|██████████| 3/3 [00:00<00:00, 14.72it/s]


Epoch 1: val_loss = 2.1375


100%|██████████| 93/93 [00:00<00:00, 128.88it/s]


Epoch 2: train_loss = 1.1137


100%|██████████| 3/3 [00:00<00:00, 21.23it/s]


Epoch 2: val_loss = 0.5693


100%|██████████| 93/93 [00:00<00:00, 153.84it/s]


Epoch 3: train_loss = 0.5668


100%|██████████| 3/3 [00:00<00:00, 21.01it/s]


Epoch 3: val_loss = 0.4894


100%|██████████| 93/93 [00:00<00:00, 156.36it/s]


Epoch 4: train_loss = 0.3979


100%|██████████| 3/3 [00:00<00:00, 18.85it/s]


Epoch 4: val_loss = 0.4389


100%|██████████| 93/93 [00:00<00:00, 154.88it/s]


Epoch 5: train_loss = 0.4169


100%|██████████| 3/3 [00:00<00:00, 20.87it/s]


Epoch 5: val_loss = 0.5511


100%|██████████| 93/93 [00:00<00:00, 151.95it/s]


Epoch 6: train_loss = 0.3788


100%|██████████| 3/3 [00:00<00:00, 20.12it/s]


Epoch 6: val_loss = 0.3358


100%|██████████| 93/93 [00:00<00:00, 153.56it/s]


Epoch 7: train_loss = 0.3510


100%|██████████| 3/3 [00:00<00:00, 20.21it/s]


Epoch 7: val_loss = 0.3237


100%|██████████| 93/93 [00:00<00:00, 155.12it/s]


Epoch 8: train_loss = 0.3187


100%|██████████| 3/3 [00:00<00:00, 18.84it/s]


Epoch 8: val_loss = 0.4141


100%|██████████| 93/93 [00:00<00:00, 152.80it/s]


Epoch 9: train_loss = 0.3734


100%|██████████| 3/3 [00:00<00:00, 20.13it/s]


Epoch 9: val_loss = 0.3092


100%|██████████| 93/93 [00:00<00:00, 149.77it/s]


Epoch 10: train_loss = 0.3610


100%|██████████| 3/3 [00:00<00:00, 21.60it/s]


Epoch 10: val_loss = 0.3042


100%|██████████| 4/4 [00:00<00:00, 26.57it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 256-128_relu_sgd_lr001_10ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/80f8ff857aa943b7bf2030d02d00c672
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 25848.974609375
COMET INFO:     MAPE            : 54.25483703613281
COMET INFO:     R2              : 0.47071486711502075
COMET INFO:     RMSE            : 39843.307493229026
COMET INFO:     train_loss [10] : (0.3187155310184725, 10.723097780699371)
COMET INFO:     val_loss [10]   : (0.3042480448881785, 2.1375258763631186)
COMET INFO:   Others:
COMET INFO:     Name         : 256-128_relu_

Epoch 1: train_loss = 10.3934


100%|██████████| 3/3 [00:00<00:00, 11.39it/s]


Epoch 1: val_loss = 1.2097


100%|██████████| 93/93 [00:01<00:00, 74.54it/s]


Epoch 2: train_loss = 1.0316


100%|██████████| 3/3 [00:00<00:00, 12.09it/s]


Epoch 2: val_loss = 0.9551


100%|██████████| 93/93 [00:01<00:00, 77.46it/s]


Epoch 3: train_loss = 0.6552


100%|██████████| 3/3 [00:00<00:00, 14.85it/s]


Epoch 3: val_loss = 0.4144


100%|██████████| 93/93 [00:00<00:00, 117.80it/s]


Epoch 4: train_loss = 0.4030


100%|██████████| 3/3 [00:00<00:00, 18.45it/s]


Epoch 4: val_loss = 0.4377


100%|██████████| 93/93 [00:00<00:00, 98.67it/s] 


Epoch 5: train_loss = 0.3722


100%|██████████| 3/3 [00:00<00:00, 17.11it/s]


Epoch 5: val_loss = 0.3442


100%|██████████| 93/93 [00:00<00:00, 117.41it/s]


Epoch 6: train_loss = 0.3345


100%|██████████| 3/3 [00:00<00:00, 19.07it/s]


Epoch 6: val_loss = 0.3528


100%|██████████| 93/93 [00:00<00:00, 115.40it/s]


Epoch 7: train_loss = 0.3298


100%|██████████| 3/3 [00:00<00:00, 18.87it/s]


Epoch 7: val_loss = 0.3637


100%|██████████| 93/93 [00:00<00:00, 115.35it/s]


Epoch 8: train_loss = 0.3656


100%|██████████| 3/3 [00:00<00:00, 18.78it/s]


Epoch 8: val_loss = 0.3401


100%|██████████| 93/93 [00:00<00:00, 111.08it/s]


Epoch 9: train_loss = 0.3183


100%|██████████| 3/3 [00:00<00:00, 18.87it/s]


Epoch 9: val_loss = 0.3266


100%|██████████| 93/93 [00:00<00:00, 115.33it/s]


Epoch 10: train_loss = 0.3036


100%|██████████| 3/3 [00:00<00:00, 18.80it/s]


Epoch 10: val_loss = 0.4309


100%|██████████| 93/93 [00:00<00:00, 103.68it/s]


Epoch 11: train_loss = 0.3189


100%|██████████| 3/3 [00:00<00:00, 18.47it/s]


Epoch 11: val_loss = 0.2966


100%|██████████| 93/93 [00:00<00:00, 114.98it/s]


Epoch 12: train_loss = 0.3181


100%|██████████| 3/3 [00:00<00:00, 18.67it/s]


Epoch 12: val_loss = 0.3472


100%|██████████| 93/93 [00:00<00:00, 113.71it/s]


Epoch 13: train_loss = 0.2849


100%|██████████| 3/3 [00:00<00:00, 13.39it/s]


Epoch 13: val_loss = 0.4988


100%|██████████| 93/93 [00:01<00:00, 78.82it/s]


Epoch 14: train_loss = 0.2780


100%|██████████| 3/3 [00:00<00:00, 12.59it/s]


Epoch 14: val_loss = 0.2660


100%|██████████| 4/4 [00:00<00:00, 14.86it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 512-256_relu_sgd_lr001_14ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/4f298b130434451e98186049c66575e6
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 23555.8671875
COMET INFO:     MAPE            : 44.31068420410156
COMET INFO:     R2              : 0.49928992986679077
COMET INFO:     RMSE            : 38752.853830395514
COMET INFO:     train_loss [14] : (0.27798823147050794, 10.393419866279888)
COMET INFO:     val_loss [14]   : (0.2660355468591054, 1.2096578280131023)
COMET INFO:   Others:
COMET INFO:     Name         : 512-256_relu_s

Epoch 1: train_loss = 4.5199


100%|██████████| 3/3 [00:00<00:00, 16.48it/s]


Epoch 1: val_loss = 0.6773


100%|██████████| 93/93 [00:00<00:00, 154.52it/s]


Epoch 2: train_loss = 0.5947


100%|██████████| 3/3 [00:00<00:00, 19.92it/s]


Epoch 2: val_loss = 0.6160


100%|██████████| 93/93 [00:00<00:00, 142.81it/s]


Epoch 3: train_loss = 0.5129


100%|██████████| 3/3 [00:00<00:00, 20.82it/s]


Epoch 3: val_loss = 0.5067


100%|██████████| 93/93 [00:00<00:00, 147.08it/s]


Epoch 4: train_loss = 0.4608


100%|██████████| 3/3 [00:00<00:00, 21.11it/s]


Epoch 4: val_loss = 0.4663


100%|██████████| 93/93 [00:00<00:00, 152.57it/s]


Epoch 5: train_loss = 0.4231


100%|██████████| 3/3 [00:00<00:00, 17.67it/s]


Epoch 5: val_loss = 0.4308


100%|██████████| 93/93 [00:00<00:00, 154.05it/s]


Epoch 6: train_loss = 0.3924


100%|██████████| 3/3 [00:00<00:00, 20.11it/s]


Epoch 6: val_loss = 0.4036


100%|██████████| 93/93 [00:00<00:00, 149.19it/s]


Epoch 7: train_loss = 0.3668


100%|██████████| 3/3 [00:00<00:00, 20.42it/s]


Epoch 7: val_loss = 0.3895


100%|██████████| 93/93 [00:00<00:00, 145.61it/s]


Epoch 8: train_loss = 0.3501


100%|██████████| 3/3 [00:00<00:00, 20.46it/s]


Epoch 8: val_loss = 0.4194


100%|██████████| 93/93 [00:00<00:00, 154.52it/s]


Epoch 9: train_loss = 0.3277


100%|██████████| 3/3 [00:00<00:00, 18.42it/s]


Epoch 9: val_loss = 0.3593


100%|██████████| 93/93 [00:00<00:00, 154.83it/s]


Epoch 10: train_loss = 0.3198


100%|██████████| 3/3 [00:00<00:00, 20.42it/s]


Epoch 10: val_loss = 0.3417


100%|██████████| 4/4 [00:00<00:00, 24.03it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 256-128_tanh_sgd_lr001_10ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/d4e2b4f70fcf424ba1b8b5aa55d5d5d2
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 26650.671875
COMET INFO:     MAPE            : 54.3856315612793
COMET INFO:     R2              : 0.4334130883216858
COMET INFO:     RMSE            : 41223.39995682064
COMET INFO:     train_loss [10] : (0.3198299814936935, 4.5199062766567355)
COMET INFO:     val_loss [10]   : (0.34166792035102844, 0.6773276527722677)
COMET INFO:   Others:
COMET INFO:     Name         : 256-128_tanh_sgd_l

Epoch 1: train_loss = 4.3344


100%|██████████| 3/3 [00:00<00:00, 18.10it/s]


Epoch 1: val_loss = 0.6780


100%|██████████| 93/93 [00:00<00:00, 112.50it/s]


Epoch 2: train_loss = 0.6382


100%|██████████| 3/3 [00:00<00:00, 17.79it/s]


Epoch 2: val_loss = 0.5430


100%|██████████| 93/93 [00:00<00:00, 113.11it/s]


Epoch 3: train_loss = 0.5290


100%|██████████| 3/3 [00:00<00:00, 17.51it/s]


Epoch 3: val_loss = 0.4705


100%|██████████| 93/93 [00:00<00:00, 116.93it/s]


Epoch 4: train_loss = 0.4440


100%|██████████| 3/3 [00:00<00:00, 16.25it/s]


Epoch 4: val_loss = 0.4825


100%|██████████| 93/93 [00:00<00:00, 114.93it/s]


Epoch 5: train_loss = 0.4122


100%|██████████| 3/3 [00:00<00:00, 17.06it/s]


Epoch 5: val_loss = 0.3962


100%|██████████| 93/93 [00:00<00:00, 111.53it/s]


Epoch 6: train_loss = 0.3751


100%|██████████| 3/3 [00:00<00:00, 16.43it/s]


Epoch 6: val_loss = 0.3748


100%|██████████| 93/93 [00:00<00:00, 111.39it/s]


Epoch 7: train_loss = 0.3584


100%|██████████| 3/3 [00:00<00:00, 17.03it/s]


Epoch 7: val_loss = 0.4273


100%|██████████| 93/93 [00:00<00:00, 113.29it/s]


Epoch 8: train_loss = 0.3437


100%|██████████| 3/3 [00:00<00:00, 17.88it/s]


Epoch 8: val_loss = 0.3557


100%|██████████| 93/93 [00:00<00:00, 114.51it/s]


Epoch 9: train_loss = 0.3339


100%|██████████| 3/3 [00:00<00:00, 16.21it/s]


Epoch 9: val_loss = 0.3516


100%|██████████| 93/93 [00:01<00:00, 72.38it/s]


Epoch 10: train_loss = 0.3479


100%|██████████| 3/3 [00:00<00:00, 11.47it/s]


Epoch 10: val_loss = 0.3861


100%|██████████| 93/93 [00:01<00:00, 72.19it/s]


Epoch 11: train_loss = 0.3547


100%|██████████| 3/3 [00:00<00:00, 11.70it/s]


Epoch 11: val_loss = 0.3592


100%|██████████| 93/93 [00:00<00:00, 113.52it/s]


Epoch 12: train_loss = 0.3567


100%|██████████| 3/3 [00:00<00:00, 18.79it/s]


Epoch 12: val_loss = 0.4049


100%|██████████| 93/93 [00:00<00:00, 110.86it/s]


Epoch 13: train_loss = 0.3307


100%|██████████| 3/3 [00:00<00:00, 18.37it/s]


Epoch 13: val_loss = 0.6562


100%|██████████| 93/93 [00:00<00:00, 113.84it/s]


Epoch 14: train_loss = 0.3359


100%|██████████| 3/3 [00:00<00:00, 18.13it/s]


Epoch 14: val_loss = 0.2974


100%|██████████| 4/4 [00:00<00:00, 22.58it/s]
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     name                  : 512-256_tanh_sgd_lr001_14ep
COMET INFO:     url                   : https://www.comet.com/gp5-team1/oskelly-mlp/f7d1579c07904530bc1ef568fa95479f
COMET INFO:   Metrics [count] (min, max):
COMET INFO:     MAE             : 24938.09765625
COMET INFO:     MAPE            : 51.57059860229492
COMET INFO:     R2              : 0.490986704826355
COMET INFO:     RMSE            : 39072.8501136019
COMET INFO:     train_loss [14] : (0.3307094096496541, 4.334428774733698)
COMET INFO:     val_loss [14]   : (0.2974245349566142, 0.677970806757609)
COMET INFO:   Others:
COMET INFO:     Name         : 512-256_tanh_sgd_lr0

In [46]:
results_df = pd.DataFrame(results)
results_df

,name,hidden_1,hidden_2,activation,optimizer,lr,epochs,MAE,RMSE,MAPE,R2
0,128-64_relu_adam_lr0001_10ep,128,64,relu,Adam,0.0010,10,29941.181641,51144.449865,60.246311,0.127881
1,256-128_relu_adam_lr0001_10ep,256,128,relu,Adam,0.0010,10,28249.083984,50759.880851,55.044510,0.140947
2,512-256_relu_adam_lr0001_10ep,512,256,relu,Adam,0.0010,10,24753.523438,42906.203934,40.475483,0.386211
3,512-256_relu_adam_lr0005_14ep,512,256,relu,Adam,0.0005,14,23981.755859,41762.341697,41.853439,0.418501
4,768-384_relu_adam_lr0005_14ep,768,384,relu,Adam,0.0005,14,25551.587891,45348.402045,47.026695,0.314349
5,256-128_tanh_adam_lr0001_10ep,256,128,tanh,Adam,0.0010,10,37926.570312,57179.446937,80.684090,-0.090081
6,512-256_tanh_adam_lr0005_14ep,512,256,tanh,Adam,0.0005,14,26925.169922,42990.130216,53.474915,0.383807
7,768-384_tanh_adam_lr0005_14ep,768,384,tanh,Adam,0.0005,14,22951.035156,37207.441836,44.247787,0.538429
8,256-128_relu_sgd_lr001_10ep,256,128,relu,SGD,0.0100,10,25848.974609,39843.307493,54.254837,0.470715
9,512-256_relu_sgd_lr001_14ep,512,256,relu,SGD,0.0100,14,23555.867188,38752.853830,44.310684,0.499290


In [47]:
best_result = results_df.sort_values("R2", ascending=False).iloc[0]
best_result

,7
name,768-384_tanh_adam_lr0005_14ep
hidden_1,768
hidden_2,384
activation,tanh
optimizer,Adam
lr,0.0005
epochs,14
MAE,22951.035156
RMSE,37207.441836
MAPE,44.247787


сохраним таблицу c теми данными, на которых проводились экспериметы

In [48]:
np.save("X_train.npy", X_train)
np.save("y_train.npy", y_train)

np.save("X_val.npy", X_val)
np.save("y_val.npy", y_val)

np.save("X_test.npy", X_test)
np.save("y_test.npy", y_test)

## выводы
В этой части были обучены несколько простых MLP-моделей

меняли в экпериментах самые базовые гиперпараметры - размер скрытых слоёв, функция активации, оптимизатор, learning rate и количество эпох

Лучший результат показала модель 768-384_tanh_adam_lr0005_14ep. У неё ошибка MAE составила примерно 22 951 рублей, а R квадрат  0.538. так же видно, что более крупные модели в целом работают лучше маленьких. Также Adam оказался стабильнее SGD